File: app/app/Service/Activity/UnlikeService.swift
import Alamofire

final class UnlikeService {

    static let shared = UnlikeService()

    func create(
        memberId: Int64,
    ) async throws {
        try await PrivateNetworkManager.shared.requestVoid(
            "/unlikes/\(memberId)",
            method: .post,
            encoding: JSONEncoding.default
        )
    }

    func delete(
        memberId: Int64,
    ) async throws {
        try await PrivateNetworkManager.shared.requestVoid(
            "/unlikes/\(memberId)",
            method: .delete,
            encoding: JSONEncoding.default
        )
    }

    func gets(
        cursorId: Int64?,
        cursorDateAt: String?,
        size: Int = 20,
    ) async throws -> CursorResponse<ActivityRowResponse> {
        var parameters: [String: Any] = [
            "size": size,
        ]

        if let cursorId {
            parameters["cursorId"] = cursorId
        }
        if let cursorDateAt {
            parameters["cursorDateAt"] = cursorDateAt
        }

        return try await PrivateNetworkManager.shared.request(
            "/unlikes",
            method: .get,
            parameters: parameters,
            encoding: URLEncoding.default,
            as: CursorResponse<ActivityRowResponse>.self
        )
    }

    func getsByRank(
        gender: String,
        cursorId: Int64?,
        cursorScore: Int64?,
        size: Int = 20,
    ) async throws -> CursorScoreResponse<RankRowResponse> {
        var parameters: [String: Any] = [
            "gender": gender,
            "size": size,
        ]

        if let cursorId {
            parameters["cursorId"] = cursorId
        }
        if let cursorScore {
            parameters["cursorScore"] = cursorScore
        }

        return try await PrivateNetworkManager.shared.request(
            "/unlikes/rank",
            method: .get,
            parameters: parameters,
            encoding: URLEncoding.default,
            as: CursorScoreResponse<RankRowResponse>.self
        )
    }
}

File: server/src/main/kotlin/com/blueoauld/server/activity/application/UnlikeService.kt
package com.blueoauld.server.activity.application

import com.blueoauld.server.activity.application.response.ActivityRowResponse
import com.blueoauld.server.activity.application.response.RankRowResponse
import com.blueoauld.server.activity.entity.Unlike
import com.blueoauld.server.activity.repository.UnlikeRepository
import com.blueoauld.server.common.dto.response.CursorResponse
import com.blueoauld.server.common.dto.response.CursorScoreResponse
import com.blueoauld.server.common.exception.CustomException
import com.blueoauld.server.common.exception.type.ErrorCode.ACTIVITY_03
import com.blueoauld.server.common.exception.type.ErrorCode.ACTIVITY_04
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Instant

@Service
class UnlikeService(

    private val unlikeRepository: UnlikeRepository
) {

    @Transactional
    fun create(memberId: Long, targetId: Long) {
        if (unlikeRepository.existsByFromIdAndToId(memberId, targetId)) {
            throw CustomException(ACTIVITY_03)
        }

        val unlike = Unlike(
            fromId = memberId,
            toId = targetId
        )
        unlikeRepository.save(unlike)
    }

    @Transactional
    fun delete(memberId: Long, targetId: Long) {
        val count = unlikeRepository.deleteByFromIdAndToId(memberId, targetId)

        if (count == 0) {
            throw CustomException(ACTIVITY_04)
        }
    }

    @Transactional(readOnly = true)
    fun gets(
        memberId: Long,
        cursorId: Long?,
        cursorDateAt: Instant?,
        size: Int
    ): CursorResponse<ActivityRowResponse> {
        val rows = unlikeRepository.findAllByCursor(
            memberId = memberId,
            cursorId = cursorId,
            cursorDateAt = cursorDateAt,
            size = size + 1
        ).map {
            ActivityRowResponse.from(it)
        }

        return CursorResponse.of(rows, size, { it.activityId }, { it.createdAt })
    }

    @Transactional(readOnly = true)
    fun getsByRank(
        gender: String,
        cursorId: Long?,
        cursorScore: Long?,
        size: Int
    ): CursorScoreResponse<RankRowResponse> {
        val rows = unlikeRepository.findAllByRank(
            gender = gender,
            cursorId = cursorId,
            cursorScore = cursorScore,
            size = size + 1
        ).map {
            RankRowResponse.from(it)
        }

        return CursorScoreResponse.of(rows, size, { it.memberId }, { it.unlikes })
    }
}